google/cloud/security/scanner/audit/bigquery_rules_engine.py

```diff
@@ -0,0 +1,284 @@
+        self.rule_index = rule_index
+        self.rules = rules
+
+    def find_policy_violations(self, bigquery_acl):
+        """Find BigQuery acl violations in the rule book.
+
+        Args:
+            bigquery_acl (BigqueryAccessControls): BigQuery ACL resource
+
+        Yields:
+            namedtuple: Returns RuleViolation named tuple
+        """
+        dataset_id_bool = True
+        special_group_bool = True
+        user_email_bool = True
+        domain_bool = True
+        group_email_bool = True
+        role_bool = True
+
+        if self.rules.dataset_id != '^.+$':
+            dataset_id_bool = re.match(self.rules.dataset_id,
+                                       bigquery_acl.dataset_id)
+        if self.rules.special_group != '^.+$':
+            special_group_bool = re.match(self.rules.special_group,
+                                          bigquery_acl.special_group)
+        if self.rules.user_email != '^.+$':
+            user_email_bool = re.match(self.rules.user_email,
+                                       bigquery_acl.user_email)
+        if self.rules.domain != '^.+$':
+            domain_bool = re.match(self.rules.domain, bigquery_acl.domain)
+        if self.rules.group_email != '^.+$':
+            group_email_bool = re.match(self.rules.group_email,
+                                        bigquery_acl.group_email)
+        if self.rules.role != '^.+$':
+            role_bool = re.match(self.rules.role, bigquery_acl.role)
+
+        should_raise_violation = (
+            (dataset_id_bool is not None and dataset_id_bool) and
+            (special_group_bool is not None and special_group_bool) and
+            (user_email_bool is not None and user_email_bool) and
+            (domain_bool is not None and domain_bool) and
+            (group_email_bool is not None and group_email_bool) and
+            (role_bool is not None and role_bool))
+
+        if should_raise_violation:
+            yield self.RuleViolation(
+                resource_type='project',
+                resource_id=bigquery_acl.project_id,
+                rule_name=self.rule_name,
+                rule_index=self.rule_index,
+                violation_type='BIGQUERY_VIOLATION',
+                dataset_id=bigquery_acl.dataset_id,
+                role=bigquery_acl.role,
+                special_group=bigquery_acl.special_group,
+                user_email=bigquery_acl.user_email,
+                domain=bigquery_acl.domain,
+                group_email=bigquery_acl.group_email)
+
+    # Rule violation.
+    # resource_type: string
+    # resource_id: string
+    # rule_name: string
+    # rule_index: int
+    # violation_type: BIGQUERY_VIOLATION
+    # dataset_id: string
+    # role: string
+    # special_group: string
+    # user_email: string
+    # domain: string
+    # group_email: string
+    RuleViolation = namedtuple('RuleViolation',
+                               ['resource_type', 'resource_id', 'rule_name',
```

Comment:
nit: Same suggestion as above. It's probably best to define this list as a frozenset at the class or module level.